### ISSUE TITLE
How to run the different model variants decribed in the paper?

### ISSUE THREAD

**Issue by lena-schwert** [intfloat/SimKGC#4]
Hello Liang,

I have a question about how you would run the different model variants using your code.

First off, based on Table 2 (see below), I guess it would be best to use SimKGC with all 3 negative types (IB+PB+SN) if I run it on a subset of Wikidata5M-Trans?

![image](https://user-images.githubusercontent.com/63639195/165340849-ef1e179d-d077-4af5-a964-6febbb6cc86a.png)

When trying to do that, I found that this would be specified by adding the corresponding flags which are described in `config.py`. So if understand correctly I need to add certain flags to the `main.py` call inside `train_wiki.sh`.

However I am not entirely sure how to do it, so could you confirm/extend what I am writing below?

From what I can see:

- to use in-batch negatives (IB): Unclear. I don't see an option for that inside the config. Is this always enabled?

- to use pre-batch negatives (PB): add the flags `--pre-batch 2 --pre-batch-weight 0.5`  as described in the hyperparameters in chapter 4.1

- to use self-negatives (SN): add the flag `--use-self-negative`

Thanks!

**Comment by intfloat** [comment]
Hi Lena,

Most of your understanding is correct:
1. In-batch negatives are always enabled.
2. Pre-batch negatives can be specified with `--pre-batch 2`, set it to 0 if you'd like to turn it off.
3. Self-negatives are enabled with `--use-self-negative`, delete this option to disable self-negatives.

Please check out the command in [scripts/train_wiki.sh](scripts/train_wiki.sh):

https://github.com/intfloat/SimKGC/blob/main/scripts/train_wiki.sh#L22-L39

By default, we set `--pre-batch 0` since it slows down training a little bit and shows only slight improvement.

**Comment by intfloat** [comment]
Glad to hear about your experimental results~

The main reason that pre-batch negatives would slow down training is due to the following code:

https://github.com/intfloat/SimKGC/blob/4388ebc0c0011fe333bc5a98d0613ab0d1825ddc/models.py#L125-L127

The `construct_mask` function runs in the main process instead of worker process, and it requires cpu-intensive nested loops to mask out false negatives.

Getting slightly better results than reported in the paper is expected.

https://github.com/intfloat/SimKGC/blob/4388ebc0c0011fe333bc5a98d0613ab0d1825ddc/trainer.py#L158-L161

For the results in the paper, we only use the first loss (given a head entity and a relation, choose the correct tail entity). After the paper got accepted, we added the second loss (given a tail entity, choose the correct <head entity, relation> tuple).

On question "Did you experiment with using more epochs on Wikidata5M? If not, do you expect that that would improve performance?"

No, I did not experiment with more epochs. I tried training a model with 10% of the data, and the results are not very far from training on the entire dataset, though I lost the exact numbers. I would expect further improvements if training for more epochs, but my guess is that the improvements are unlikely to be very substantial (I may be wrong).

Thank you!

Liang
